### DISCUSSION TITLE
How to disable OpenTelemetry

### DISCUSSION THREAD

**Discussion by mkaszub** [open-telemetry/opentelemetry-dotnet#5627] [Q&A] [answered]
Hi,
how can I control if open telemetry is enabled/disabled based on the configuration. What I want to achieve is that in some environments my metrics should not be collected in my API controllers. 

I configure OpenTelemetry metrics with auto-start using the AddOpenTelemetry extension from OpenTelemetry.Extensions.Hosting.

```
services.AddOpenTelemetry().ConfigureResource(builder => builder
                    .AddService(serviceName: "some-service"))
                .WithMetrics(builder => builder
                        .AddMeter("my_meter")
                        .AddOtlpExporter((otlpOptions, metricReaderOptions) =>
                            {
                                otlpOptions.Endpoint = new Uri("http://some_url:4317");
                                otlpOptions.Protocol = OtlpExportProtocol.Grpc;
                            }
                        )
                );
```
Later in my controller's constructor I'm creating meter:
`_meter = new Meter("my_meter", "1.0.0");`

... and finally in one of the function I'm collecting metric
```
var countHistogram = _meter.CreateHistogram<int>("count");
...
countHistogram.Record(<some time>);
```

What would be the best approach to disable metrics collection? I was reading in spec about [No-Op]([url](https://opentelemetry.io/docs/specs/otel/metrics/noop/)) but I cannot find it in the library.

**Comment by cijothomas** [comment, marked as answer]
All metrics (traces too) are explicit opt-in - i.e metrics must be enabled by using AddMeter() API. If you want to only enable metrics in one environment, but not in another, the easiest way is by controlling AddMeter() calls. For example, you can keep the list of meter names in config file (like appsettings.json), and call AddMeter() using the names from the config. By controlling the config for each environment, you  can control the metrics itself.

if (app.Environment.IsDevelopment())
{
 
}
else
{
AddMeter("My-meter");
}

**Reply by mkaszub** [reply]
Thanks @cijothomas . What about piece of the code that I have in controller.
```
var countHistogram = _meter.CreateHistogram<int>("count");
...
countHistogram.Record(<some time>);
```
Is it safe to call it when meter disabled? Would be great to not have "if" around all my counters/histograms.

**Reply by cijothomas** [reply]
it is absolutely fine! If the Meter is not added to the MeterProvider, then everything will "no-ops", and you dont need to do anything extra.

**Reply by egbert-algra** [reply]
How to disable a meter temporarily, i.e. I am looking for an option to 'meter.Disable()' and 'meter.Enable()' from code

**Reply by cijothomas** [reply]
I don't think there is disable/enable. You can Dispose() a meter to stop it, but it can't be "activated" again. I think you can create a new Meter with same name and it might do the trick.. (To be tested. I am not super sure on this)

**Reply by egbert-algra** [reply]
Thanks @cijothomas ! 

I tested this 'method' as work around and found the following to work:
Configuration:
1. configure the meter builder with the meter name(s) in scope, and build the meter provider.
Metrics 'push':
1. create a new meter with the name previously used during meterProvider build
2. create instrument (e.g. meter.CreateGauge) and record one or more value
3. dispose the meter
4. ForceFlush the meter provider.
(strangely enough, the step 3) dispose is needed before the step 4) ForceFlush, to prevent duplicates when repeating the procedure with 

This recipe works in case of one wanting to 'push' subsets of metrics with specifically recorded values.

Using otel exporter version 1.15.0 and dotnet framework 4.8.1

Update: Instruments cannot be re-created with the same name at random from meters - there is a limit of 1000.
See https://github.com/open-telemetry/opentelemetry-dotnet/issues/5950#issuecomment-2490077049
